lib/IDE/PostfixCompletion.cpp

```diff
@@ -189,23 +189,24 @@ void PostfixCompletionCallback::sawSolutionImpl(
 
   bool BaseIsStaticMetaType = S.isStaticallyDerivedMetatype(ParsedExpr);
 
+  bool ExpectsNonVoid = false;
   SmallVector<Type, 4> ExpectedTypes;
   if (ExpectedTy) {
     ExpectedTypes.push_back(ExpectedTy);
-  }
-
-  bool ExpectsNonVoid = false;
-  ExpectsNonVoid |= ExpectedTy && !ExpectedTy->isVoid();
-  ExpectsNonVoid |=
-      !ParentExpr && CS.getContextualTypePurpose(CompletionExpr) != CTP_Unused;
+    ExpectsNonVoid = !ExpectedTy->isVoid();
+  } else {
+    ExpectsNonVoid |= !ParentExpr &&
```

Comment:
I don’t know when it would happen but if expected type is non-void but it’s unused by the constraint system, we should still have `ExpectsNonVoid = false`, I think. Did you intentionally change it? Same for the check in the loop below.

Reply:
> if expected type is non-void but it’s unused by the constraint system, we should still have `ExpectsNonVoid = false`, I think

Changing this means that for e.g the following:

```
  let _ = MyStruct4 {
    1
  } #^INIT_FALLBACK_1^#
// INIT_FALLBACK_1: Begin completions, 3 items
// INIT_FALLBACK_1-DAG: Keyword[self]/CurrNominal:          .self[#MyStruct4<Int>#]; name=self
// INIT_FALLBACK_1-DAG: Decl[InstanceMethod]/CurrNominal/TypeRelation[Invalid]: .testStructMethod()[#Void#]; name=testStructMethod()
```

`testStructMethod()` is no longer marked as having an `Invalid` type relation, which I guess is reasonable? It's not invalid, although it's also not particularly useful either.

Reply:
Yes, I think that’s reasonable. If `ExpectedTy` is null, we don’t know what the expected type is and thus we can’t tell that it’s not `Void`.

What I was talking about, though, is that `ExpectsNonVoid` should be `false` if `ExpectedTy` is eg. `Int` but the contextual type purpose of the code completion expression is unused.

Reply:
How's this?

Reply:
Almost. 

```cpp
  auto isResultUsed = [&]() {
    if (!ParentExpr &&
        CS.getContextualTypePurpose(CompletionExpr) != CTP_Unused) {
      return true;
    }
    for (auto SAT : S.targets) {
      if (SAT.second.getAsExpr() == CompletionExpr)
        return SAT.second.getExprContextualTypePurpose() != CTP_Unused;
    }
    // No contextual type purpose, assume it's unused.
    return false;
  };

  bool ExpectsNonVoid = (ExpectedTy && !ExpectedTy->isVoid()) || isResultUsed();
```

That also maintains the current behavior.